File: cpp/6-arp.cpp
//dont use this. wont work

#include <iostream>
#include <cstring>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <net/if_arp.h>
#include <net/if.h>
#include <sys/ioctl.h>
#include <arpa/inet.h>

void send_arp_request(const char* ifname, const char* target_ip) {
    int sockfd = socket(AF_INET, SOCK_PACKET, htons(ETH_P_ARP));
    if (sockfd < 0) {
        perror("Socket creation failed");
        return;
    }

    struct ifreq ifr;
    strcpy(ifr.ifr_name, ifname);
    if (ioctl(sockfd, SIOCGIFHWADDR, &ifr) < 0) {
        perror("ioctl");
        close(sockfd);
        return;
    }

    struct arpreq arp;
    memset(&arp, 0, sizeof(arp));
    struct sockaddr_in* sin = (struct sockaddr_in*)&arp.arp_pa;
    sin->sin_family = AF_INET;
    sin->sin_addr.s_addr = inet_addr(target_ip);
    sin = (struct sockaddr_in*)&arp.arp_ha;
    sin->sin_family = ARPHRD_ETHER;
    memcpy(arp.arp_ha.sa_data, ifr.ifr_hwaddr.sa_data, 6);
    arp.arp_hln = 6;
    arp.arp_pln = 4;
    arp.arp_op = ARPOP_REQUEST;

    if (ioctl(sockfd, SIOCSARP, &arp) < 0) {
        perror("ioctl");
        close(sockfd);
        return;
    }

    std::cout << "ARP request sent" << std::endl;

    close(sockfd);
}

void send_rarp_request(const char* ifname, const char* target_mac) {
    // RARP is obsolete and not widely supported, so we won't implement it here
    std::cout << "RARP is obsolete and not widely supported." << std::endl;
}

int main() {
    const char* ifname = "eth0"; // Change to your interface name
    const char* target_ip = "192.168.1.1"; // Change to your target IP address

    send_arp_request(ifname, target_ip);

    return 0;
}

File: python/6-arp.py
import socket
import struct
import binascii

def send_arp_request(iface, target_ip):
    # Create a raw socket
    sock = socket.socket(socket.AF_PACKET, socket.SOCK_RAW)

    # Get source MAC address
    src_mac = hex(get_mac_address(iface)).lstrip('0x')
    src_mac = src_mac[:2] + ':' + src_mac[2:4] + ':' + src_mac[4:6] + ':' + src_mac[6:8] + ':' + src_mac[8:10] + ':' + src_mac[10:12]

    # Ethernet header
    ether_header = struct.pack('!6s6s2s', binascii.unhexlify('ff'*6), binascii.unhexlify(src_mac.replace(':', '')), b'\x08\x06')

    # ARP header
    arp_header = struct.pack('!HHBBH6s4s6s4s', 1, 0x0800, 6, 4, 1, binascii.unhexlify(src_mac.replace(':', '')),
                             socket.inet_aton('0.0.0.0'), binascii.unhexlify('00'*6), socket.inet_aton(target_ip))

    # Send ARP request
    sock.send(ether_header + arp_header)

def send_rarp_request(iface, target_mac):
    # RARP is obsolete and not widely supported, so we won't implement it here
    print("RARP is obsolete and not widely supported.")

def get_mac_address(iface):
    # Get the MAC address of the given interface
    try:
        with open(f'/sys/class/net/{iface}/address') as f:
            mac = f.read().strip()
            return int(mac.replace(':', ''), 16)
    except IOError:
        print("Error: Interface not found or unable to read MAC address.")
        exit(1)

def main():
    iface = input("Enter the interface name: ")
    target_ip = input("Enter the target IP address for ARP request: ")
    send_arp_request(iface, target_ip)

    target_mac = input("Enter the target MAC address for RARP request: ")
    send_rarp_request(iface, target_mac)

if __name__ == "__main__":
    main()
